core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java

```diff
@@ -0,0 +1,68 @@
+package org.springframework.security.access.hierarchicalroles;
+
+import org.springframework.util.Assert;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.springframework.security.access.hierarchicalroles.RoleHierarchyUtils.roleHierarchyFromMap;
+
+/**
+ * Class to provide a fluent API to build a {@link String} representing a {@link RoleHierarchy}.
+ * Relies on  {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)} to build the string.
+ */
+public class RoleHierarchyBuilder {
+	private final Map<String, List<String>> roleBranches;
+
+	private RoleHierarchyBuilder() {
+		this.roleBranches = new LinkedHashMap<>();
+	}
+
+	public static RoleHierarchyBuilder newBuilder() {
+		return new RoleHierarchyBuilder();
+	}
+
+	/**
+	 * Create a new hierarchy branch to define the hierarchy.
+	 * @param role that will inherit other roles.
+	 * @return a branch builder to define hierarchy of the <code>role</code>.
+	 * @throws IllegalArgumentException if <code>role</code> is null or empty.
+	 */
+	public RoleBranchBuilder role(String role) {
+		Assert.hasText(role, "role must not be empty");
+		return new RoleBranchBuilder(role, this);
+	}
+
+	/**
+	 * Builds the hierarchy using {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)}
+	 * @return a string representation of a role hierarchy.
+	 */
+	public String build() {
+		return roleHierarchyFromMap(roleBranches);
+	}
+
+	public static class RoleBranchBuilder {
+		private final RoleHierarchyBuilder parentBuilder;
+		private final String role;
+
+		RoleBranchBuilder(String role, RoleHierarchyBuilder roleHierarchyBuilder) {
+			this.role = role;
+			this.parentBuilder = roleHierarchyBuilder;
+
+		}
+
+		/**
+		 * Implied role(s).
+		 * @param impliedRoles role name(s) implied by the role.
+		 * @return the same {@link RoleHierarchyBuilder}
+		 * @throws IllegalArgumentException if <code>impliedRoles</code> is null or empty
+		 * or contains any null element.
+		 */
+		public RoleHierarchyBuilder inheritsFromAllOf(String... impliedRoles) {
```

Comment:
Can we call this `implies`? I think the simpler name makes the DSL more readable. It also aligns the method name with the parameter name and JavaDoc.